### DISCUSSION TITLE
How do I use a liquid variable inside a liquid variable?

### DISCUSSION THREAD

**Discussion by arpeasal** [11ty/eleventy#2321] [Q&A] [answered]
How do I use a liquid variable inside a liquid variable?

Here's what I'm trying to do in my eleventy blog.

Using pagination, I create a page for each category that lists all of the posts in that category.  I have this working fine.

What I want to do now is add a description for the category on each category page.

In the `_data` directory I have created a sub-directory for each category.  In each of those directories I have created a file, `description.json`.  
In each of those files I have an object:
```
{
  "portugues": "Esta é a descrição da categoria vocabulario em portugues.",
  "ingles": "This is the vocabulary category description in English."
}
```

(Of course, the value part in the key:value pairs is different in each file.)

I tested this by using one example in the categories pagination page.  I used `{{ vocabulary.description.ingles }}`.  On each category page, where I placed this variable, it produced a line "This is the vocabulary category description in English.", as expected.  So the structure of the data is correct.

What I need now is a way to get the pagination to use the `{{ category }}` variable innstead of "vocabulary" in `{{vocabulary.description.ingles }}`.

I tried `{{ {{ category }}.description.ingles }}` and I get the error:
`TemplateContentRenderError was thrown
[11ty] > unexpected token at "{{ categoria", file:./source/category/index.md, line:3, col:4`

After reading some suggestions when I did a search for "How do I use a liquid variable inside a liquid variable?" I tried: 
`{{ [ category ].description.ingles }}`
I got the error message: 
`TemplateContentRenderError was thrown
[11ty] > unexpected token at "categoria].de...", file:./source/categoria/index.md, line:3, col:4`

I also tried: `{% {{categoria}}.descricao.portugues %}`
and `{% {{ categoria }}.descricao.portugues %}`
and `{% {%categoria%}.descricao.portugues %}`
and `{{ {%categoria%}.descricao.portugues }}`
and `{{ {% categoria %}.descricao.portugues }}`
None of those worked either.

So is there a way to use a liquid variable inside a liquid variable?

Or am I going about this all wrong.  Is there a better way to insert a description for each category on that categories page?

**Comment by pdehaan** [comment, marked as answer]
I think the solution is to add ANOTHER nested directory in your _data/ directory.
Here's my layout:

```sh
$ tree

.
└── src/
    ├── _data/
    │   └── categories/
    │       └── foo/
    │           └── description.json
    └── index.liquid

4 directories, 2 files
```
(Where "/categories/" is my new directory, and "/foo/" is my category name.)

Now, in my src/index.liquid file I can do this:
```liquid
---
title: 11ty-2321
key: foo
---

e: {{ categories[key].description.ingles }}
p: {{ categories[key].description.portugues }}
```

### OUTPUT
```html
e: This is the vocabulary category description in English.
p: Esta é a descrição da categoria vocabulario em portugues.
```

**Reply by arpeasal** [reply]
Thanks for the reply.
But unfortunately, I'm noy really following this.
First, mysrc/index.liquid file is my home page and on that page I want a list of the most recent posts with a snippet/excerpt of each.  If I'm seeing what you've written correctly, that home page would have descriptions of the categories.  That is not what I want.

What I want is this.  When you go to a post page, under the title of the post there is a listing (horizontally) of the categories that the post is in.  The names of the categories are clickable links that link to a page for that category that lists all posts that have that category.  For example, if I have a post on fruits, one of the categories for this post would be vocabulary.  On the post page, if you click on the link for the vocabulary category, it takes you to a page called Vocabulary and all of the posts on the site that have the category "vocabulary" are listed on that page with links to the actual posts.  This is already working.

What I want is to add near the top of the Category page, a description of what that category is about.  Again, I don't want this information on the home page, just on each category page.  

Second, I have no idea what the significance of "title: 11ty-2321" is.

And third (and last), I'm not sure what you mean "and "/foo/" is my category name".  Do you mean that in the src/_data/categories directory I should make a separate directory for each category?  I'm confused.

**Reply by pdehaan** [reply]
Sorry, maybe I'm misunderstanding something here.

> **In the _data directory I have created a sub-directory for each category.** In each of those directories I have created a file,  description.json.
> In each of those files I have an object:
> ```json
> {
>   "portugues": "Esta é a descrição da categoria vocabulario em portugues.",
>   "ingles": "This is the vocabulary category description in English."
> }
> ```

Same, I just called my category "foo". I interpreted your comment above as you had a `_data/{{ category }}/description.json` file for each category.

But I think you're close. Using `{{ [ category ].description.ingles }}` isn't valid Liquid syntax, but by moving the data files into a subfolder, you can use `{{ categories[ category ].description.ingles }}` which is valid and should show the localized text you want (assuming you change your folder structure from "_data/vocabulary/description.json" to "_data/categories/vocabulary/description.json"). But you don't have to call the folder "categories", you can call it anything you want.

And the `title: 11ty-2321` front matter has no significance at all. it was just the bug/discussion number here.

**Reply by arpeasal** [reply]
Note, In this program, I use categorias instead of categories and categoria instead of category.

Well, something funky is going on.  All I did was, as you suggested, make a sub-directory source/_data/categorias and move my three category directories into that sub directory.  I immediately got an error message:
```
[11ty] Problem writing Eleventy templates: (more in DEBUG output)
[11ty] > cats.forEach is not a function

`TypeError` was thrown:
[11ty]     TypeError: cats.forEach is not a function
        at /galpao/backup_apenas_local/websites/eleventy-alternative/antigos/projeto10-changed_include_to_render/.eleventy.js:32:9
        at Array.forEach (<anonymous>)
```
 I have no idea what is causing this error, but if I remove the categorias sub-directory, the error goes away and the site works fine, except for not displaying the category description.  And this is the case whether I completely remove the categorias sub-directory and all its contents or if I keep its contents, but in the _data directory. 

I've included a screenshot of the file structure and the code of the source/categoria/index.md file.  Note that the line "{{ categorias[ categoria ].description.ingles }}" which is immediately after the front matter is the line you suggested and is just there for testing.  It does NOT appear to have anything to do the error as the error appears even if I remove that line from this file.

Again, the rest of the stuff works fine, so I don't know what it is that's breaking it.

![eleventy-screenshot](https://user-images.githubusercontent.com/102467466/162844075-0a60e4e3-012d-459a-8ddf-79ba031be3a3.png)

**Reply by pdehaan** [reply]
I'm not sure. I wonder if it's a conflict between `collections.categorias` and `_data/categorias/`, but that doesn't exactly make sense.

And `cats.forEach is not a function` sounds like there is a `cats` array somewhere around .eleventy.js:32:9 which is probably getting a `null` or `undefined` value for `cats`.

Do you have a public repo somewhere that I can try cloning and building locally?
Otherwise, see if choosing a different name for the _data/categorias/ directory and see if that makes a difference (if it still fails, we can at least eliminate the fact that it's a weird naming issue).

**Reply by pdehaan** [reply]
Cant tell if I'm reading the indentation correctly in the screenshot, but is that a source/_data/categoria/site.json file, or is that under source/_data/site.json?

**Reply by pdehaan** [reply]
For what it's worth, this worked in my minimal test case: https://github.com/pdehaan/11ty-2321

Also, per your screenshot above, I think you need to change `description` to `descricao` here:

```diff
-{{ categories[ categoria ].description.ingles }}
+{{ categorias[ categoria ].descricao.ingles }}
```

**Reply by arpeasal** [reply]
Well, it must have been a weird naming issue.

The problem was solved when I chose a different name.  I changed "categorias" to "bobby", so the vocabulario directory is now in source/_data/bobby/vocabulario.  As I said, the error disappeared.  Also "{{ bobby[ categoria ].description.ingles }}" in the categoria/index.md file displays the category description as I want.

This may not be relevant now, but, no, I don't have a public repo.  That's down my list after I get this initially working.  I do have a server (a vps) that I can probably upload the file to.  I just zipped the entire project folder, so that includes EVERYTHING, including the node_modules.  It is about 21.6MB.  If you want, tomorrow I can upload it to my server and send you a link to download it.  But I don't know why you'd want me to.

Regarding the site.json file, it is source/_data/site.json, NOT source/_data/categoria/site.json file.

Well, I'm done for the night.  Thanks again for your help.

**Reply by pdehaan** [reply]
Odd. Not sure where the conflict would be w/ naming a folder `categorias` but sounds like a naming conflict with some other variables.
Nope, you don't need to ZIP up the site for me anymore. But you can close the discussion if it answered your question.

**Reply by arpeasal** [reply]
Not sure how to "close the discussion".  I marked it as answered.  Do I need to do something else to "close" it?
